### PR TITLE
ISO TX data path including ISOAL

### PR DESCRIPTION
**Pull request by mtpr-ot**
Bluetooth: controller: ISO TX data path including ISOAL
- ISO TX data path for HCI and support for vendor path
- ISO-AL segmentation of framed PDUs
- Insertion of segment headers
- Reconstruction and storing of CIG reference point in ULL
- Calculation and insertion of of Time-Offset
- Exit error spooling in ISO-AL on detecting start
- ISO-AL TX unframed fragmentation

tests: bluetooth: Updates for ISOAL unittest
Minor adjustments due to implementation changes.

Bluetooth: host: Set ISO type for CIS at request
Set BT_ISO_CHAN_TYPE_CONNECTED as type for connected ISO at CIS_REQ.

### CODE REVIEW

**Review by niag-Oticon — CHANGES_REQUESTED** [review]
Minor deletion.

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 1]
```diff
@@ -679,6 +725,11 @@ static isoal_status_t isoal_rx_framed_consume(struct isoal_sink *sink,
 			timeoffset = seg_hdr->timeoffset;
 			anchorpoint = meta->timestamp;
 			latency = session->latency_framed;
+			/* TODO: This needs to be rechecked.
+			 * timeoffset seems to be in us
+			 * latency seems to be in us
+			 * anchorpoint seems to be in ticks (16 = 1us)
+			 */
```
[the remark this answers is not in the document]
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/isoal.c]: Done

**Comment by mtpr-ot** [comment]
@niag-Oticon, do you want to update with recent ISOAL fixes?

File: subsys/bluetooth/controller/ll_sw/ull_iso_types.h [hunk 2]
```diff
@@ -25,6 +25,18 @@
...
+#define IS_ADV_HANDLE(_handle) \
+	(((_handle) >= BT_CTLR_ADV_ISO_STREAM_HANDLE_BASE) && \
+	((_handle) <= (BT_CTLR_ADV_ISO_STREAM_HANDLE_BASE + BT_CTLR_ADV_ISO_STREAM_MAX - 1)))
+#else
+#define IS_ADV_HANDLE(_handle) 0
+#endif /* CONFIG_BT_CTLR_ADV_ISO */
+
+/* ISO LL conformance tests require a PDU size of maximum 251 bytes */
+#define ISO_TX_BUFFER_SIZE 251
+
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_iso_types.h:39]:
```suggestion
```

Revert this change and put back `CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE` as LE Read Buffer Size command allows a range of 1 to 0xFFFF for ISO_Data_Packet_Length

File: subsys/bluetooth/controller/Kconfig [hunk 3]
```diff
@@ -215,15 +215,6 @@ config BT_CTLR_ISO_TX_BUFFERS
...
 
-config BT_CTLR_ISO_TX_BUFFER_SIZE
-	int "Isochronous Tx buffer size"
-	depends on BT_CTLR_ADV_ISO || BT_CTLR_CONN_ISO
-	range 1 4095
-	default 27
-	help
-	  Size of the Isochronous Tx buffers and the value returned in HCI LE
-	  Read Buffer Size V2 command response.
-
```
Comment (cvinayak) [inline subsys/bluetooth/controller/Kconfig:226]: Revert this back, as this is the value used to return ISO_Data_Packet_Length  in HCI LE Read Buffer Size v2 command response

File: samples/bluetooth/hci_rpmsg/nrf5340_cpunet_iso_broadcast.conf [hunk 4]
```diff
@@ -31,7 +31,6 @@ CONFIG_BT_CTLR_ADV_PERIODIC=y
 CONFIG_BT_CTLR_ADV_ISO=y
 CONFIG_BT_CTLR_ADV_ISO_PDU_LEN_MAX=251
 CONFIG_BT_CTLR_ISO_TX_BUFFERS=3
-CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE=251
```
Comment (cvinayak) [inline samples/bluetooth/hci_rpmsg/nrf5340_cpunet_iso_broadcast.conf:34]: Add back

File: subsys/bluetooth/controller/hci/hci.c [hunk 5]
```diff
@@ -1344,7 +1348,7 @@ static void le_read_buffer_size_v2(struct net_buf *buf, struct net_buf **evt)
 
 	rp->acl_max_len = sys_cpu_to_le16(LL_LENGTH_OCTETS_TX_MAX);
 	rp->acl_max_num = CONFIG_BT_BUF_ACL_TX_COUNT;
-	rp->iso_max_len = sys_cpu_to_le16(CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE);
+	rp->iso_max_len = sys_cpu_to_le16(ISO_TX_BUFFER_SIZE);
```
Comment (cvinayak) [inline subsys/bluetooth/controller/hci/hci.c:1351]:
```suggestion
	rp->iso_max_len = sys_cpu_to_le16(CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE);
```

File: subsys/bluetooth/controller/hci/hci.c [hunk 6]
```diff
@@ -5149,113 +5159,191 @@ int hci_iso_handle(struct net_buf *buf, struct net_buf **evt)
...
+	 *    higher layer (Not implemented)
+	 * -- Any other method of determining Time_Offset (Not implemented)
+	 */
+	sdu_frag_tx.time_stamp = HAL_TICKER_TICKS_TO_US(ticker_ticks_now_get());
+	if (ts_flag) {
+		/* Overwrite time stamp with HCI provided time stamp */
+		time_stamp = net_buf_pull_mem(buf, sizeof(*time_stamp));
+		len -= sizeof(*time_stamp);
+		sdu_frag_tx.time_stamp = *time_stamp;
+	}
```
Comment (cvinayak) [inline subsys/bluetooth/controller/hci/hci.c:5186]:
```suggestion
	if (ts_flag) {
		/* Overwrite time stamp with HCI provided time stamp */
		time_stamp = net_buf_pull_mem(buf, sizeof(*time_stamp));
		len -= sizeof(*time_stamp);
		sdu_frag_tx.time_stamp = *time_stamp;
	} else {
		sdu_frag_tx.time_stamp = HAL_TICKER_TICKS_TO_US(ticker_ticks_now_get());
	}
```

File: subsys/bluetooth/controller/hci/hci.c [hunk 7]
```diff
@@ -5149,113 +5159,191 @@ int hci_iso_handle(struct net_buf *buf, struct net_buf **evt)
...
+
+	/* Extract ISO data header if included (PB_Flag 0b00 or 0b10) */
+	sdu_frag_tx.packet_sn = 0;
+	sdu_frag_tx.iso_sdu_length = 0;
+	if ((pb_flag & 0x01) == 0) {
+		iso_data_hdr = net_buf_pull_mem(buf, sizeof(*iso_data_hdr));
+		len -= sizeof(*iso_data_hdr);
+		sdu_frag_tx.packet_sn = iso_data_hdr->sn;
+		sdu_frag_tx.iso_sdu_length = iso_data_hdr->slen;
+	}
```
Comment (cvinayak) [inline subsys/bluetooth/controller/hci/hci.c:5196]:
```suggestion
	if ((pb_flag & 0x01) == 0) {
		iso_data_hdr = net_buf_pull_mem(buf, sizeof(*iso_data_hdr));
		len -= sizeof(*iso_data_hdr);
		sdu_frag_tx.packet_sn = iso_data_hdr->sn;
		sdu_frag_tx.iso_sdu_length = iso_data_hdr->slen;
	} else {
		sdu_frag_tx.packet_sn = 0;
		sdu_frag_tx.iso_sdu_length = 0;
	}
```

File: subsys/bluetooth/controller/hci/hci_driver.c [hunk 8]
```diff
@@ -111,7 +111,7 @@ isoal_status_t sink_sdu_emit_hci(const struct isoal_sink         *sink_ctx,
 {
 	struct bt_hci_iso_ts_data_hdr *data_hdr;
 	uint16_t packet_status_flag;
-	uint16_t slen, slen_packed;
+	uint16_t slen_packed;
 	struct bt_hci_iso_hdr *hdr;
 	uint16_t handle_packed;
 	struct net_buf *buf;
```
Comment (cvinayak) [inline subsys/bluetooth/controller/hci/hci_driver.c:117]:
```suggestion
	struct bt_hci_iso_hdr *hdr;
	uint16_t handle_packed;
	uint16_t slen_packed;
	struct net_buf *buf;
```

File: subsys/bluetooth/controller/ll_sw/pdu.h [hunk 9]
```diff
@@ -688,25 +688,19 @@ struct pdu_data_llctrl_cte_rsp {
...
+	uint8_t  p_phy;
+	/* c_max_sdu:12
+	 * rfu:3
+	 * framed:1
+	 */
+	uint8_t  c_max_sdu_packed[2];
+	/* p_max_sdu:12
+	 * rfu:4
+	 */
+	uint8_t  p_max_sdu[2];
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/pdu.h:703]: Any reason to remove bitfields?

File: subsys/bluetooth/controller/ll_sw/ull_conn_iso.c [hunk 10]
```diff
@@ -88,8 +103,10 @@ struct ll_conn_iso_stream *ll_conn_iso_stream_acquire(void)
 {
 	struct ll_conn_iso_stream *cis = mem_acquire(&cis_free);
 
-	cis->hdr.datapath_in = NULL;
-	cis->hdr.datapath_out = NULL;
+	if (cis) {
+		cis->hdr.datapath_in = NULL;
+		cis->hdr.datapath_out = NULL;
+	}
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_conn_iso.c:109]:
```suggestion
	if (cis) {
		cis->hdr.datapath_in = NULL;
		cis->hdr.datapath_out = NULL;
	}

```

**Review by cvinayak — CHANGES_REQUESTED** [review]
Quick pass through, isoal.c not reviewed yet.

**Comment by cvinayak** [comment]
Please fix build errors compiling iso_broadcast sample and runtime missing num completes.

**Comment by cvinayak** [comment]
@mtpr-ot I have pushed my fixes as a new commit in this PR.

File: subsys/bluetooth/controller/ll_sw/pdu.h (hunk 9 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ll_sw/pdu.h:703]: Yes, the bitfields were not portable for BE architecture, for certain compilers. This is due to the 12-bit field. I don't have the technical details, but there was no way to get the compiler to align correctly.

**Review by cvinayak — DISMISSED** [review]
quick-ack.

**Comment by mtpr-ot** [comment]
> quick-ack.

@cvinayak did you manage to run your own tests with this?

**Comment by cvinayak** [comment]
> > quick-ack.
> 
> @cvinayak did you manage to run your own tests with this?

@mtpr-ot I ran tests when I send the additional commit in this PR. Not the latest, I can do now and revert back if there are any issues.

**Comment by cvinayak** [comment]
> > > quick-ack.
> > 
> > 
> > @cvinayak did you manage to run your own tests with this?
> 
> @mtpr-ot I ran tests when I send the additional commit in this PR. Not the latest, I can do now and revert back if there are any issues.

No regressions.

**Comment by mtpr-ot** [comment]
@wopu-ot, would you please review? Thanks.

File: subsys/bluetooth/controller/ll_sw/isoal.c [hunk 11]
```diff
@@ -564,6 +594,20 @@ static isoal_status_t isoal_rx_unframed_consume(struct isoal_sink *sink,
 		sp->sdu_status |= ISOAL_SDU_STATUS_LOST_DATA;
 	}
 
+	/* BT Core V5.3 : Vol 4 HCI I/F : Part G HCI Func. Spec.:
+	 * 5.4.5 HCI ISO Data packets
+	 * If Packet_Status_Flag equals 0b10 then PB_Flag shall equal 0b10.
+	 * When Packet_Status_Flag is set to 0b10 in packets from the Controller to the
+	 * Host, there is no data and ISO_SDU_Length shall be set to zero.
+	 *
+	 * TODO: Move to hci_driver to allow vendow path to have dedicated handling.
```
Comment (wopu-ot) [inline subsys/bluetooth/controller/ll_sw/isoal.c:603]: vendow -> vendor

**Review by wopu-ot — CHANGES_REQUESTED** [review]
There are some cosmetic changes I would like to see, but no fundamental objections from me.

File: subsys/bluetooth/controller/ll_sw/isoal.h [hunk 12]
```diff
@@ -148,6 +148,7 @@ struct isoal_sdu_produced {
 struct isoal_pdu_produced {
 	/** Contents information provided at PDU allocation */
 	struct isoal_pdu_buffer contents;
+	uint64_t payload_number;
```
[the remark this answers is not in the document]
Comment (niag-Oticon) [reply subsys/bluetooth/controller/ll_sw/isoal.h:151]: Nice catch! It does not appear to be used any longer and should be removed.

**Review by wopu-ot — APPROVED** [review]

**Review by cvinayak — APPROVED** [review]

**Comment by mtpr-ot** [comment]
Are additional review approvals necessary?
